impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java

```diff
@@ -116,92 +156,105 @@ public ExpiryPolicyBuilder access(Duration access) {
+    return new BaseExpiryPolicy<>(create, access, update);
   }
 
   /**
    * Simple implementation of the {@link ExpiryPolicy} interface allowing to set constants to each expiry types.
    */
-  private static class BaseExpiryPolicy implements ExpiryPolicy<Object, Object> {
+  private static class BaseExpiryPolicy<K, V> implements ExpiryPolicy<K, V> {
 
-    private final Duration create;
-    private final Duration access;
-    private final Duration update;
+    private final BiFunction<? super K, ? super V, Duration> create;
+    private final BiFunction<? super K, ? super Supplier<? extends V>, Duration> access;
+    private final TriFunction<? super K, ? super Supplier<? extends V>, ? super V, Duration> update;
 
-    BaseExpiryPolicy(Duration create, Duration access, Duration update) {
+    BaseExpiryPolicy(BiFunction<? super K, ? super V, Duration> create,
+                     BiFunction<? super K, ? super Supplier<? extends V>, Duration> access,
+                     TriFunction<? super K, ? super Supplier<? extends V>, ? super V, Duration> update) {
       this.create = create;
       this.access = access;
       this.update = update;
     }
-    @Override
-    public Duration getExpiryForCreation(Object key, Object value) {
-      return create;
-    }
 
     @Override
-    public Duration getExpiryForAccess(Object key, Supplier<? extends Object> value) {
-      return access;
+    public Duration getExpiryForCreation(K key, V value) {
+      return create.apply(key, value);
     }
 
     @Override
-    public Duration getExpiryForUpdate(Object key, Supplier<? extends Object> oldValue, Object newValue) {
-      return update;
+    public Duration getExpiryForAccess(K key, Supplier<? extends V> value) {
+      return access.apply(key, value);
     }
 
     @Override
-    public boolean equals(final Object o) {
-      if (this == o) return true;
-      if (o == null || getClass() != o.getClass()) return false;
-
-      final BaseExpiryPolicy that = (BaseExpiryPolicy) o;
+    public Duration getExpiryForUpdate(K key, Supplier<? extends V> oldValue, V newValue) {
+      return update.apply(key, oldValue, newValue);
+    }
+  }
 
-      if (!Objects.equals(access, that.access)) return false;
-      if (!Objects.equals(create, that.create)) return false;
-      if (!Objects.equals(update, that.update)) return false;
+  private static final class TimeToLiveExpiryPolicy extends BaseExpiryPolicy<Object, Object> {
+    private final Duration ttl;
 
-      return true;
+    TimeToLiveExpiryPolicy(Duration ttl) {
+      super(
+        (a, b) -> ttl,
+        (a, b) -> null,
+        (a, b, c) -> ttl);
+      this.ttl = ttl;
     }
 
     @Override
     public int hashCode() {
-      int result = Objects.hashCode(create);
-      result = 31 * result + Objects.hashCode(access);
-      result = 31 * result + Objects.hashCode(update);
-      return result;
+      return ttl.hashCode();
     }
 
     @Override
-    public String toString() {
```

Comment:
We've lost this `toString()` implementation... should we add it back (at least for the TTL and TTI subclasses)?

Reply:
Yes. Why not now that you have added attributes for TTL and TTI.